File: app/src/main/java/io/github/marcocipriani01/telescopetouch/control/SensorOrientationController.java
/*
 * Copyright 2021 Marco Cipriani (@marcocipriani01)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.marcocipriani01.telescopetouch.control;

import android.content.SharedPreferences;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.util.Log;

import javax.inject.Inject;

import io.github.marcocipriani01.telescopetouch.TelescopeTouchApp;

import static io.github.marcocipriani01.telescopetouch.ApplicationConstants.DISABLE_GYRO_PREF;
import static io.github.marcocipriani01.telescopetouch.ApplicationConstants.SKY_MAP_HIGH_REFRESH_PREF;

/**
 * Sets the direction of view from the orientation sensors.
 *
 * @author John Taylor
 */
public class SensorOrientationController extends AbstractController implements SensorEventListener {

    private final static String TAG = TelescopeTouchApp.getTag(SensorOrientationController.class);
    private final SensorManager manager;
    private final Sensor rotationSensor;
    private final Sensor geomagneticRotationSensor;
    private final SharedPreferences preferences;

    @Inject
    SensorOrientationController(SensorManager manager, SharedPreferences preferences) {
        this.manager = manager;
        this.rotationSensor = manager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
        this.geomagneticRotationSensor = manager.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR);
        this.preferences = preferences;
    }

    @Override
    public void start() {
        if (manager != null) {
            if (preferences.getBoolean(DISABLE_GYRO_PREF, false)) {
                Log.d(TAG, "Using geomagnetic rotation sensor");
                manager.registerListener(this, geomagneticRotationSensor, SensorManager.SENSOR_DELAY_FASTEST);
            } else {
                Log.d(TAG, "Using gyroscope sensor");
                manager.registerListener(this, rotationSensor,
                        preferences.getBoolean(SKY_MAP_HIGH_REFRESH_PREF, false) ? SensorManager.SENSOR_DELAY_FASTEST : SensorManager.SENSOR_DELAY_GAME);
            }
        }
        Log.d(TAG, "Registered sensor listener");
    }

    @Override
    public void stop() {
        manager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        int type = event.sensor.getType();
        if (type == Sensor.TYPE_ROTATION_VECTOR) {
            model.setPhoneSensorValues(event.values);
        } else if (type == Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR) {
            model.setPhoneSensorValues(event.values);
        } else {
            Log.w(TAG, "Unknown Sensor readings");
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }
}

File: app/src/main/java/com/google/android/stardroid/control/SensorOrientationController.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.control

import android.content.SharedPreferences
import android.hardware.*
import android.util.Log
import com.google.android.stardroid.ApplicationConstants
import com.google.android.stardroid.control.SensorOrientationController
import com.google.android.stardroid.util.MiscUtil.getTag
import com.google.android.stardroid.util.smoothers.ExponentiallyWeightedSmoother
import com.google.android.stardroid.util.smoothers.PlainSmootherModelAdaptor
import javax.inject.Inject
import javax.inject.Provider

/**
 * Sets the direction of view from the orientation sensors.
 *
 * @author John Taylor
 */
class SensorOrientationController @Inject internal constructor(
    modelAdaptorProvider: Provider<PlainSmootherModelAdaptor>,
    manager: SensorManager, sharedPreferences: SharedPreferences
) : AbstractController(), SensorEventListener {
    // TODO(johntaylor): this class needs to be refactored to use the new
    // sensor API and to behave properly when sensors are not available.
    private class SensorDampingSettings(var damping: Float, var exponent: Int)

    private val manager: SensorManager?
    private var accelerometerSmoother: SensorListener? = null
    private var compassSmoother: SensorListener? = null
    private val modelAdaptorProvider: Provider<PlainSmootherModelAdaptor>
    private val rotationSensor: Sensor
    private val sharedPreferences: SharedPreferences
    override fun start() {
        val modelAdaptor = modelAdaptorProvider.get()
        if (manager != null) {
            if (!sharedPreferences.getBoolean(
                    ApplicationConstants.SHARED_PREFERENCE_DISABLE_GYRO,
                    false
                )
            ) {
                Log.d(TAG, "Using rotation sensor")
                manager.registerListener(this, rotationSensor, SensorManager.SENSOR_DELAY_GAME)
            } else {
                // TODO(jontayler): remove this code once enough it's used in few enough phones.
                Log.d(TAG, "Using classic sensors")
                Log.d(TAG, "Exponentially weighted smoothers used")
                val dampingPreference = sharedPreferences.getString(
                    ApplicationConstants.SENSOR_DAMPING_PREF_KEY,
                    ApplicationConstants.SENSOR_DAMPING_STANDARD
                )
                val speedPreference = sharedPreferences.getString(
                    ApplicationConstants.SENSOR_SPEED_PREF_KEY,
                    ApplicationConstants.SENSOR_SPEED_STANDARD
                )
                Log.d(TAG, "Sensor damping preference $dampingPreference")
                Log.d(TAG, "Sensor speed preference $speedPreference")
                var dampingIndex = 0
                if (ApplicationConstants.SENSOR_DAMPING_HIGH == dampingPreference) {
                    dampingIndex = 1
                } else if (ApplicationConstants.SENSOR_DAMPING_EXTRA_HIGH == dampingPreference) {
                    dampingIndex = 2
                } else if (ApplicationConstants.SENSOR_DAMPING_REALLY_HIGH == dampingPreference) {
                    dampingIndex = 3
                }
                var sensorSpeed = SensorManager.SENSOR_DELAY_GAME
                if (ApplicationConstants.SENSOR_SPEED_SLOW == speedPreference) {
                    sensorSpeed = SensorManager.SENSOR_DELAY_NORMAL
                } else if (ApplicationConstants.SENSOR_SPEED_HIGH == speedPreference) {
                    sensorSpeed = SensorManager.SENSOR_DELAY_FASTEST
                }
                accelerometerSmoother = ExponentiallyWeightedSmoother(
                    modelAdaptor,
                    ACC_DAMPING_SETTINGS[dampingIndex].damping,
                    ACC_DAMPING_SETTINGS[dampingIndex].exponent
                )
                compassSmoother = ExponentiallyWeightedSmoother(
                    modelAdaptor,
                    MAG_DAMPING_SETTINGS[dampingIndex].damping,
                    MAG_DAMPING_SETTINGS[dampingIndex].exponent
                )
                manager.registerListener(
                    accelerometerSmoother,
                    SensorManager.SENSOR_ACCELEROMETER,
                    sensorSpeed
                )
                manager.registerListener(
                    compassSmoother,
                    SensorManager.SENSOR_MAGNETIC_FIELD,
                    sensorSpeed
                )
            }
        }
        Log.d(TAG, "Registered sensor listener")
    }

    override fun stop() {
        Log.d(
            TAG, "Unregistering sensor listeners: " + accelerometerSmoother + ", "
                    + compassSmoother + ", " + this
        )
        manager!!.unregisterListener(accelerometerSmoother)
        manager.unregisterListener(compassSmoother)
        manager.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor != rotationSensor) {
            return
        }
        model.setPhoneSensorValues(event.values)
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        // Ignore
    }

    companion object {
        private val TAG = getTag(SensorOrientationController::class.java)

        /**
         * Parameters that control the smoothing of the accelerometer and
         * magnetic sensors.
         */
        private val ACC_DAMPING_SETTINGS = arrayOf(
            SensorDampingSettings(0.7f, 3),
            SensorDampingSettings(0.7f, 3),
            SensorDampingSettings(0.1f, 3),
            SensorDampingSettings(0.1f, 3)
        )
        private val MAG_DAMPING_SETTINGS = arrayOf(
            SensorDampingSettings(0.05f, 3),  // Derived for the Nexus One
            SensorDampingSettings(0.001f, 4),  // Derived for the unpatched MyTouch Slide
            SensorDampingSettings(0.0001f, 5),  // Just guessed for Nexus 6
            SensorDampingSettings(0.000001f, 5) // Just guessed for Nexus 6
        )
    }

    init {
        this.manager = manager
        this.modelAdaptorProvider = modelAdaptorProvider
        rotationSensor = manager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
        this.sharedPreferences = sharedPreferences
    }
}